Assets/MixedRealityToolkit/_Core/Utilities/ClippingPrimitive.cs

```diff
@@ -0,0 +1,256 @@
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Core.Utilities
+{
+    /// <summary>
+    /// An abstract primitive component to animate and visualize a clipping primitive that can be 
+    /// used to drive per pixel based clipping.
+    /// </summary>
+    public abstract class ClippingPrimitive : MonoBehaviour
+    {
+        [Tooltip("The renderer(s) that should be affected by the primitive.")]
+        [SerializeField]
+        protected List<Renderer> renderers = new List<Renderer>();
+
+        public enum Side
+        {
+            Inside = 1,
+            Outside = -1
+        }
+
+        [Tooltip("Which side of the primitive to clip pixels against.")]
+        [SerializeField]
+        protected Side clippingSide = Side.Inside;
+
+        /// <summary>
+        /// The renderer(s) that should be affected by the primitive.
+        /// </summary>
+        public Side ClippingSide => clippingSide;
+
+        protected abstract string Keyword { get; }
+        protected abstract string KeywordProperty { get; }
+        protected abstract string ClippingSideProperty { get; }
+
+        protected MaterialPropertyBlock materialPropertyBlock;
+        protected Dictionary<Material, bool> modifiedMaterials = new Dictionary<Material, bool>();
+        protected List<Material> allocatedMaterials = new List<Material>();
+
+        private int clippingSideID;
+
+        public void AddRenderer(Renderer renderer)
+        {
+            if (renderer != null && !renderers.Contains(renderer))
+            {
+                renderers.Add(renderer);
+
+                Material material = GetMaterial(renderer, false);
+
+                if (material != null)
+                {
+                    ToggleClippingFeature(material, true);
+                }
+            }
+        }
+
+        public void RemoveRenderer(Renderer renderer)
+        {
+            if (renderers.Contains(renderer))
+            {
+                Material material = GetMaterial(renderer, false);
+
+                if (material != null)
+                {
+                    ToggleClippingFeature(material, false);
+                }
+
+                renderers.Remove(renderer);
+            }
+        }
+
+        protected void OnValidate()
+        {
+            ToggleClippingFeature(true);
+            RestoreUnassignedMaterials();
+        }
+
+        protected void OnEnable()
+        {
+            Initialize();
+            UpdateRenderers();
+            ToggleClippingFeature(true);
+        }
+
+        protected void OnDisable()
+        {
+            UpdateRenderers();
+            ToggleClippingFeature(false);
+        }
+
+        protected void Update()
+        {
+            if (Application.isPlaying || !Application.isEditor)
+            {
+                return;
+            }
+
+            Initialize();
+            UpdateRenderers();
+        }
+
+        protected void LateUpdate()
+        {
+            UpdateRenderers();
```

Comment:
Do we also want to check is playing or is not the editor in this update loop as well?

Reply:
Do we want both the update loops running at the same time?

Reply:
I agree this has got somewhat confusing since the preprocessor removals when moving to vNext. (This code was pushed down from ClipPlane.cs) I'll add a comment to make it a bit more clear.